### PR TITLE
util/codegen, cmd/cloner: organize and simplify cloner

### PR DESCRIPTION
**Pull request by josharian**
- util/codegen: add package
- util/codegen: add AssertStructUnchanged
- cmd/cloner: add Clone context to regen struct assignments
- cmd/cloner: simplify code
- cmd/cloner: delete some debug code
- util/codegen: add NamedTypes
- cmd/cloner: unify switch cases
- util/codegen: add ContainsPointers

### CODE REVIEW

**Comment by josharian** [comment]
I am in the process of hacking up a cmd/viewer, similar to cmd/cloner, but that generates readonly views rather than copies. It shares a lot of code and structure with cmd/cloner. This PR is preliminary work on cloner, cleaning up the code a bit and extracting some obviously re-usable parts.

Review commit-by-commit for best effect.

**Comment by dsnet** [comment]
I'm starting review now. Just wanted to comment that a few weeks ago I was exploring how to make the cloner logic nicer with the use of generics. I'd love to work with you on seeing in we can get that integrated in the future.

**Comment by josharian** [comment]
> Just wanted to comment that a few weeks ago I was exploring how to make the cloner logic nicer with the use of generics.

Cool, SGTM.

I'd also like to make the logic a bit less one-off: Instead of panicking on pointers-to-pointers, it should do recursive codegen and (as a consequence) handle a broader swath of types. Sounds scary, but in practice probably quite straightforward. But not urgent, either.

Happy to work with you on either/both of those efforts. I'll put up a WIP PR for cmd/viewer once this one is in so you can see what I'm up to there.

File: util/codegen/codegen.go [hunk 1]
```diff
@@ -38,3 +40,42 @@ func WriteFormatted(code []byte, path string) error {
...
 	return nil
 }
+
+// AssertStructUnchanged generates code that asserts at compile time that type t is unchanged.
+// tname is the named type corresponding to t.
+// ctx is a single-word context for this assertion, such as "Clone".
+// thisPkg is the package containing t.
+// If non-nil, AssertStructUnchanged will add elements to imports
+// for each package path that the caller must import for the returned code to compile.
+func AssertStructUnchanged(t *types.Struct, tname, ctx string, thisPkg *types.Package, imports map[string]struct{}) []byte {
```
Comment (dsnet) [inline util/codegen/codegen.go:80]: Optional:  Perhaps `thisPkg` should come right between `t` to match the fact that we usually refer to qualified type as `PkgPath.TypeName`

**Review by dsnet — APPROVED** [review]
Sigh... GitHub makes it hard to review these. It's all mechanical changes.

**Comment by dsnet** [comment]
My apologies for the delay, some drama with the kiddos ;)

File: util/codegen/codegen.go (hunk 1 above)
Comment (josharian) [reply util/codegen/codegen.go]: Done. Did it as an additional commit to avoid pointless merge pain.

**Pull request merged**